### PR TITLE
[RM Ch5] Load Balancer Requirements

### LINKED ISSUE #2906: [RM Ch5] 5.1.5. General PaaS Services Requirements - Load Balancer (LB)
This issue is for the specification of the LB requirements

Umbrella Issue #2785

### PR DESCRIPTION
**Pull request by pgoyal01**
Fixes #2906

### CODE REVIEW

File: doc/ref_model/chapters/chapter05.rst
Comment (rgstori) [inline doc/ref_model/chapters/chapter05.rst:197]: I would suggest to have L4 as mandatory and L7 as optional.
in Kubernetes,  Ingress controllers (L7) depends on a L4 LB anyway, so wouldn't exist without it as a standalone L7 LB entity

```suggestion
  - The Load Balancer must support OSI Level 3/4 load-balancing and may support Level 7 Load Balancing.
```

**Comment by walterkozlowski** [comment]
Typos in:
Platform Services - Load Balancer REquirements
The tbale below specifies a set of requirements for the Load Balances platform service.

@pgoyal01: can we somehow view the requirements in a table format? Any idea how to do this?  Reviewing the requirements in a current format is difficult.

**Review by walterkozlowski — APPROVED** [review]

File: doc/ref_model/chapters/chapter05.rst
Comment (karinesevilla) [inline doc/ref_model/chapters/chapter05.rst:199]: @pgoyal01 Typo -> requirements
The linewrap must be 120 characters

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [inline doc/ref_model/chapters/chapter05.rst:173]: This should underline "Requirements" also in the title.

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [inline doc/ref_model/chapters/chapter05.rst:183]: I believe you need to use grid tables for multi line cells.

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [inline doc/ref_model/chapters/chapter05.rst:186]: Please do not use raw html, like <br>. You can enforce new lines with the [block syntax](https://docutils.sourceforge.io/docs/user/rst/quickref.html#line-blocks) as it is explained [here](https://stackoverflow.com/questions/13504595/multi-line-table-cell-in-restructuredtext).

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [inline doc/ref_model/chapters/chapter05.rst:184]: What is the meaning of this in technical terms?

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [inline doc/ref_model/chapters/chapter05.rst:205]: Are there any open source solutions implementing these requirements?

**Review by CsatariGergely — CHANGES_REQUESTED** [review]

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:183]: @CsatariGergely In the preview it renders properly

**Review by karinesevilla — APPROVED** [review]

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:186]: @CsatariGergely Thanks - Done. Please review.

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:173]: @CsatariGergely Thanks - Done

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:184]: @CsatariGergely Made a minor edit -- please review. Thanks

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [inline doc/ref_model/chapters/chapter05.rst:177]: - Grid between the rows is missing
- Cell separator missing from the first column in the 2nd line of pas.lb.007
- Cell separator is missing from the first column in pas.lb.013 - pas.lb.017
- Can you please wrap it to 120 chars?

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [reply doc/ref_model/chapters/chapter05.rst]: Now it reads: `The Load Balancer should support resiliency in the  environment` I think it is either redundant to pas.lb.003 or needs more clarification.

**Review by CsatariGergely — CHANGES_REQUESTED** [review]

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:205]: @CsatariGergely Not every requirement but then these are not mandatory requirements. The Previous section lists the services and examples of open-source sw.

Comment (walterkozlowski) [reply doc/ref_model/chapters/chapter05.rst:205]: @CsatariGergely, are you satisfied with the reposne form @pgoyal01 ?  If yes, please approve.  Otherwise, let us discuss what to do.

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst]: @CsatariGergely Done

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst]: @CsatariGergely One is about resources and the other for the environment.

File: doc/ref_model/chapters/chapter05.rst
Comment (rgstori) [inline doc/ref_model/chapters/chapter05.rst:207]: typo
```suggestion
| pas.lb.010 | The Load Balancer must stateful TCP load-balancing   |                                             |
```

File: doc/ref_model/chapters/chapter05.rst
Comment (ulikleber) [inline doc/ref_model/chapters/chapter05.rst:186]: "Define local environment" reads like a requirement on the load balancer ;-).

File: doc/ref_model/chapters/chapter05.rst
Comment (ulikleber) [inline doc/ref_model/chapters/chapter05.rst:189]: Which environment now?

File: doc/ref_model/chapters/chapter05.rst
Comment (ulikleber) [inline doc/ref_model/chapters/chapter05.rst:214]: Typo as in pas.lb.010
```suggestion
| pas.lb.013 | The Load Balancer should support stateful SCTP       |                                             |
```

File: doc/ref_model/chapters/chapter05.rst
Comment (ulikleber) [inline doc/ref_model/chapters/chapter05.rst:217]: Typo as in pas.lb.010
```suggestion
| pas.lb.014 | The Load Balancer should support stateful M-TCP      |                                             |
```

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [reply doc/ref_model/chapters/chapter05.rst:189]: I think a word is missing from here.

File: doc/ref_model/chapters/chapter05.rst
Comment (CsatariGergely) [reply doc/ref_model/chapters/chapter05.rst:229]: I really think that Anuket should define requirements implementable using open source components. I do not think to be usefull to require something (even with should) in one document and define exceptions in an other one. 

By checking Octavia and NGINX as references in open source I think the following requirements are not supported in open source: 

- stateful TCP load-balancing (only NGINX supports it) 
- stateful M-TCP
- HTTP3
- QUIC

I suggest to remove these requirements.

Also Level 7 is too generic in pas.lb.015, the requirement should be about http and https

**Comment by walterkozlowski** [comment]
@CsatariGergely and @rgstori, can you please re-review the current version, and approve if you can.  We can have more discussion on these technologies after Moselle.

**Comment by walterkozlowski** [comment]
@ulikleber and @rgstor, please check whether your comments have been addressed, and approve if you can. Thanks.

**Comment by walterkozlowski** [comment]
@ulikleber, @rgstori and @pgoyal01 - can you please re-review and focus on closing anything outstanding so I could merge?  Thanks.

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:180]: @rgstori Workload -- being corrected.

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:186]: @ulikleber It is

File: doc/ref_model/chapters/chapter05.rst
Comment (pgoyal01) [reply doc/ref_model/chapters/chapter05.rst:189]: Duplicate of 003 -- deleted

**Review by ulikleber — APPROVED** [review]
Please still correct the typo state-full. We always need a full state in stateful applications ;-)

**Comment by pgoyal01** [comment]
@walterkozlowski Fixed the conflicts -- arose because of another merge that preceded.

**Comment by wmk-admin** [comment]
Merged this PR as per approved reviews of the contents and the fact that formatting issues and file conflicts have been addressed.

**Comment by CsatariGergely** [comment]
@walterkozlowski I'm pretty surprised that this pr was merged, as [my comment](https://github.com/cntt-n/CNTT/pull/2907#discussion_r872314174) was not addressed or approved. The topic was also debated with @joaoarodrigues in the RM meeting without resolution. I think we should not merge content without reaching a consensus.

**Comment by pgoyal01** [comment]
> @walterkozlowski I'm pretty surprised that this pr was merged, as [my comment](https://github.com/cntt-n/CNTT/pull/2907#discussion_r872314174) was not addressed or approved. The topic was also debated with @joaoarodrigues in the RM meeting without resolution. I think we should not merge content without reaching a consensus.

@CsatariGergely The topic was discussed in the meeting and a decision was made to proceed with the PR content as is. There was no support from anyone else for the position that Jao and you are espousing.  While every attempt has to be made to get unanimous consent, that may not always be possible. In this case, a discussion was held in the RM meeting and a decision was made. @walterkozlowski  please add your viewpoint.

**Comment by walterkozlowski** [comment]
Few facts about this PR:
1.  I have merged it after receiving 3 approvals in github
2. There was no change requested in github by @CsatariGergely 
3. We had a resolution in the RM meeting when we decided to change these requirements to "may" from "should";there was no objection to it.  And why should there by ANY?
4. I asked @CsatariGergely and others  for re-review 18 days ago; I have received responses from others, but no response from @CsatariGergely 
So from the above it is clear that  there was plenty of opportunities to discuss and express opinions.  
Anuket, and in particular Reference Model stream of work, are open for new ideas, and new evolving new standards and technologies, as this is demanded by the operator members.  And we definitely intend to maintain this position. So the arguments that some particular technology is not currently supported by a particular vendor or a particular open source project are not valid.  For specific, technical  (not "generalised, political") concerns related to any specific technology or standard, we invite to our weekly RM meetings, where we often discuss things like that.  There will be plenty of opportunities before the next release.

**Pull request merged**
